File: src/domain/style.ts
import {
  createToyBaseObject,
  createToyBaseObjectV2,
  type BaseObject,
  type BaseObjectV2,
  type IsoDateTime,
  type StableId,
} from "./common";
import type { ContextTaskType } from "./context-pack";

export interface StyleProfile extends BaseObject {
  type: "style-profile";
  ownerRef: StableId;
  scope: string;
  styleDimensions: string[];
  voiceConstraints: string[];
  doNotImitate: string[];
  evidenceRefIds: StableId[];
  transferRuleIds: StableId[];
  misuseRisks: string[];
  taskUseHints: string[];
}

export interface NarrativeDNA extends BaseObjectV2 {
  type: "narrative-dna";
  workId: StableId;
  readerExperienceGoal: string;
  promiseProgressPayoffPattern: string;
  tensionEngine: string;
  sceneRhythm: string;
  signatureMoves: string[];
  evidenceRefIds: StableId[];
  coverage: string;
  confidence: "low" | "medium" | "high";
  mechanismRefIds: StableId[];
  transferRuleRefs: StableId[];
  applicableTaskTypes: ContextTaskType[];
  applicableScope: string;
  misuseRisks: string[];
  doNotCopy: string[];
  prohibitedResidue: string[];
  currentWorkConflictChecks: string[];
}

export interface MicroTechnique extends BaseObject {
  type: "micro-technique";
  name: string;
  mechanism: string;
  useCase: string;
  constraints: string[];
  sourceEvidenceRefs: StableId[];
  transferRuleIds: StableId[];
}

export interface ImageryRecord extends BaseObject {
  type: "imagery-record";
  workId: StableId;
  imageLabel: string;
  meaning: string;
  firstUseRef: StableId;
  recurrenceRules: string[];
  currentState: string;
}

export interface BookMechanic extends BaseObject {
  type: "book-mechanic";
  ownerRef: StableId;
  mechanicName: string;
  mechanicPurpose: string;
  operatingRules: string[];
  failureModes: string[];
  evidenceRefIds: StableId[];
}

export function createToyStyleProfile(timestamp: IsoDateTime): StyleProfile {
  return {
    ...createToyBaseObject(
      "style-profile",
      "style-profile-toy",
      "Toy Practical Pressure Style",
      timestamp,
      "active",
    ),
    ownerRef: "work-public-toy",
    scope: "single invented scene",
    styleDimensions: ["plain pressure", "visible cost"],
    voiceConstraints: ["Use concrete cause and effect."],
    doNotImitate: ["Do not copy source phrasing."],
    evidenceRefIds: ["evidence-toy-pressure"],
    transferRuleIds: ["transfer-rule-toy"],
    misuseRisks: ["Pressure is stated after the action."],
    taskUseHints: ["scene-planning", "style-revision"],
  };
}

export function createToyNarrativeDNA(timestamp: IsoDateTime): NarrativeDNA {
  return {
    ...createToyBaseObjectV2(
      "narrative-dna",
      "narrative-dna-toy",
      "Toy Narrative DNA",
      timestamp,
      "active",
    ),
    workId: "work-public-toy",
    readerExperienceGoal: "Make choices feel visibly costly.",
    promiseProgressPayoffPattern: "Promise a cost, escalate it, pay it off with a choice.",
    tensionEngine: "Visible scarcity before commitment.",
    sceneRhythm: "Set cost, force choice, show consequence.",
    signatureMoves: ["cost-before-choice"],
    evidenceRefIds: ["evidence-toy-pressure"],
    coverage: "single invented scene",
    confidence: "medium",
    mechanismRefIds: ["reference-mechanism-toy"],
    transferRuleRefs: ["transfer-rule-toy"],
    applicableTaskTypes: ["outline-generation", "draft-generation"],
    applicableScope: "original scenes that need practical pressure",
    misuseRisks: ["Using scarcity after the choice weakens causality."],
    doNotCopy: ["Do not copy source scenes."],
    prohibitedResidue: ["No copied names", "No copied event sequence"],
    currentWorkConflictChecks: ["Skip if the current work already resolved the cost."],
  };
}

File: crates/core/src/model/style.rs
use serde::{Deserialize, Serialize};

use super::common::{BaseObject, StableId};
use super::context_pack::ContextTaskType;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StyleProfile {
    #[serde(flatten)]
    pub base: BaseObject,
    pub owner_ref: StableId,
    pub scope: String,
    pub style_dimensions: Vec<String>,
    pub voice_constraints: Vec<String>,
    pub do_not_imitate: Vec<String>,
    pub evidence_ref_ids: Vec<StableId>,
    pub transfer_rule_ids: Vec<StableId>,
    pub misuse_risks: Vec<String>,
    pub task_use_hints: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NarrativeDNA {
    #[serde(flatten)]
    pub base: BaseObject,
    pub work_id: StableId,
    pub reader_experience_goal: String,
    pub promise_progress_payoff_pattern: String,
    pub tension_engine: String,
    pub scene_rhythm: String,
    pub signature_moves: Vec<String>,
    pub evidence_ref_ids: Vec<StableId>,
    pub coverage: String,
    pub confidence: String,
    pub mechanism_ref_ids: Vec<StableId>,
    pub transfer_rule_refs: Vec<StableId>,
    pub applicable_task_types: Vec<ContextTaskType>,
    pub applicable_scope: String,
    pub misuse_risks: Vec<String>,
    pub do_not_copy: Vec<String>,
    pub prohibited_residue: Vec<String>,
    pub current_work_conflict_checks: Vec<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MicroTechnique {
    #[serde(flatten)]
    pub base: BaseObject,
    pub name: String,
    pub mechanism: String,
    pub use_case: String,
    pub constraints: Vec<String>,
    pub source_evidence_refs: Vec<StableId>,
    pub transfer_rule_ids: Vec<StableId>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ImageryRecord {
    #[serde(flatten)]
    pub base: BaseObject,
    pub work_id: StableId,
    pub image_label: String,
    pub meaning: String,
    pub first_use_ref: StableId,
    pub recurrence_rules: Vec<String>,
    pub current_state: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BookMechanic {
    #[serde(flatten)]
    pub base: BaseObject,
    pub owner_ref: StableId,
    pub mechanic_name: String,
    pub mechanic_purpose: String,
    pub operating_rules: Vec<String>,
    pub failure_modes: Vec<String>,
    pub evidence_ref_ids: Vec<StableId>,
}
